src/main/kotlin/com/github/hanseter/json/editor/SchemaNormalizer.kt

```diff
@@ -69,12 +69,25 @@ object SchemaNormalizer {
                     resolutionScope, schema
             )
         } else {
-            resolveRefs(
-                    resolveRefFromUrl(ref, resolutionScope).use {
-                        JSONObject(JSONTokener(it.reader(Charsets.UTF_8)))
-                    },
-                    resolutionScope
-            )
+            val resolvedSchema = resolveRefFromUrl(ref, resolutionScope)
+
+            val resolvedFragment = resolvedSchema.inputStream.use {
+                val fullObject = JSONObject(JSONTokener(it.reader(Charsets.UTF_8)))
+
+                if (!resolvedSchema.fragment.isNullOrBlank()) {
+                    val queryResult: Any? = fullObject.optQuery(resolvedSchema.fragment)
+
+                    if (queryResult !is JSONObject) {
+                        throw java.lang.IllegalArgumentException("Target of fragment ${resolvedSchema.fragment} is not an object")
+                    }
+
+                    queryResult to fullObject
+                } else {
+                    fullObject to fullObject
+                }
+            }
+
+            resolveRefs(resolvedFragment.first, resolvedSchema.location, resolvedFragment.second)
```

Comment:
```suggestion
            val fullObject = resolvedSchema.inputStream.use {
                JSONObject(JSONTokener(it.reader(Charsets.UTF_8)))
            }
            val resolvedFragment = if (!resolvedSchema.fragment.isNullOrBlank()) {
                    val queryResult: Any? = fullObject.optQuery(resolvedSchema.fragment)

                    if (queryResult !is JSONObject) {
                        throw IllegalArgumentException("Target of fragment ${resolvedSchema.fragment} is not an object")
                    }

                    queryResult
                } else {
                    fullObject
                }

            resolveRefs(resolvedFragment, resolvedSchema.location, fullObject)
```
Maybe that would also work and would get rid of the tuple and reduce nesting by 1?

Reply:
Done. I didn't use the suggestion directly because I realized I could remove the temporary `queryResult` variable completely.